### PR TITLE
Faster ASE NL filtering

### PR DESCRIPTION
**Pull request by frostedoyster**
Speeds up the ASE calculator by speeding up the NL filtering.
With the new changes, the filtering is negligible compared to the time it takes to build the NL (with vesin).

# Contributor (creator of pull-request) checklist

 ~~- [ ] Tests updated (for new features and bugfixes)?~~
 ~~- [ ] Documentation updated (for new features)?~~
 ~~- [ ] Issue referenced (for PRs that solve an issue)?~~

# Reviewer checklist

 ~~- [ ] CHANGELOG updated with public API or any other important changes?~~

----
 📚 [Download documentation preview for this pull-request](https://nightly.link/metatensor/metatensor/actions/artifacts/2111701974.zip)

### CODE REVIEW

**Comment by ceriottm** [comment]
Does this also apply to LAMMPS-metatensor?

**Comment by frostedoyster** [comment]
No, unfortunately this only applies to the ASE interface

**Comment by frostedoyster** [comment]
@Luthaf docs look broken, I don't think it's the changes in this PR

**Comment by Luthaf** [comment]
The changes looks good, could you share some benchmark results? It would be nice to always do this when doing "this is faster" PR =)

**Comment by frostedoyster** [comment]
Using this script and with respect to the old timings from `master`:
```
import metatensor_lj_test
import torch
from metatensor.torch.atomistic.ase_calculator import MetatensorCalculator


CUTOFF = 5.0
SIGMA = 1.5808
EPSILON = 0.1729


model = metatensor_lj_test.lennard_jones_model(
    atomic_type=28,
    cutoff=CUTOFF,
    sigma=SIGMA,
    epsilon=EPSILON,
    length_unit="Angstrom",
    energy_unit="eV",
    with_extension=False,
)

model.save("model.pt")
device = "cpu"
calculator = MetatensorCalculator("model.pt", device=device)


from ase.build import bulk

atoms = bulk("Ni", cubic=True)
atoms *= (8, 8, 8)
print("Number of atoms:", atoms.get_number_of_atoms())
atoms.set_calculator(calculator)

from ase.md import VelocityVerlet
from ase.md.velocitydistribution import MaxwellBoltzmannDistribution

MaxwellBoltzmannDistribution(atoms, temperature_K=300)

dyn = VelocityVerlet(atoms, 0.5)

import time

dyn.run(10)

start = time.time()
dyn.run(100)
end = time.time()

print("Time taken for 100 steps: ", end - start, "s")
```

The improvements are
CPU: `3.65` s to `2.37 s`
GPU: `3.00 s` to `2.05 s`

File: python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py [hunk 1]
```diff
@@ -503,28 +503,32 @@ def _compute_ase_neighbors(atoms, options, dtype, device):
             cutoff=options.engine_cutoff(engine_length_unit="angstrom"),
         )
```
Comment (Luthaf) [inline python/metatensor-torch/metatensor/torch/atomistic/ase_calculator.py:505]: Maybe add a small explanation about why the code is written this way for the next person coming and trying to improve readability by converting it back to a loop?

**Review by Luthaf — APPROVED** [review]

**Pull request merged**